File: backend/app/schemas/market_data.py
"""
Schemas for market data refresh and quote responses.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


QuoteSource = Literal["finnhub", "cache", "mock"]


class PriceRefreshQuoteRead(BaseModel):
    """One successfully updated quote."""

    ticker: str
    current_price: float
    provider: str
    quote_source: QuoteSource
    fetched_at: datetime


class PriceRefreshFailureRead(BaseModel):
    """One ticker that did not yield a usable price."""

    ticker: str
    reason: str


class PortfolioPriceRefreshResponse(BaseModel):
    """Summary of POST .../refresh-prices."""

    portfolio_id: int
    tickers_requested: list[str]
    updated_count: int
    failed_count: int
    skipped_count: int
    updated_quotes: list[PriceRefreshQuoteRead]
    failures: list[PriceRefreshFailureRead]
    notes: list[str] = Field(default_factory=list)
    cache_hit_count: int = 0
    provider_call_count: int = 0


class RefreshAllPricesResponse(BaseModel):
    """Summary of POST .../refresh-all-prices."""

    portfolios_processed: int
    unique_tickers_requested: list[str]
    updated_holdings_count: int
    provider_call_count: int
    cache_hit_count: int
    failed_count: int
    failures: list[PriceRefreshFailureRead]
    notes: list[str] = Field(default_factory=list)


class MarketQuoteRead(BaseModel):
    """Single-ticker quote for GET .../quote/{ticker} (no persistence)."""

    ticker: str
    current_price: float
    provider: str
    quote_source: QuoteSource
    fetched_at: datetime

File: frontend/src/api/marketData.ts
import { apiRequest } from "./client";

const MARKET_DATA_BASE = "/api/v1/market-data";

export interface PriceRefreshQuote {
  ticker: string;
  current_price: number;
  provider: string;
  quote_source: "finnhub" | "cache" | "mock";
  fetched_at: string;
}

export interface PriceRefreshFailure {
  ticker: string;
  reason: string;
}

export interface PortfolioPriceRefreshResponse {
  portfolio_id: number;
  tickers_requested: string[];
  updated_count: number;
  failed_count: number;
  skipped_count: number;
  updated_quotes: PriceRefreshQuote[];
  failures: PriceRefreshFailure[];
  notes: string[];
  cache_hit_count: number;
  provider_call_count: number;
}

export interface RefreshAllPricesResponse {
  portfolios_processed: number;
  unique_tickers_requested: string[];
  updated_holdings_count: number;
  provider_call_count: number;
  cache_hit_count: number;
  failed_count: number;
  failures: PriceRefreshFailure[];
  notes: string[];
}

export interface SymbolSearchResult {
  symbol: string;
  description: string;
  display_symbol: string;
  type: string;
  provider: string;
}

/** Ask backend to fetch Finnhub quotes and persist Holding.current_price. */
export function refreshPortfolioPrices(
  portfolioId: number,
): Promise<PortfolioPriceRefreshResponse> {
  return apiRequest<PortfolioPriceRefreshResponse>(
    `${MARKET_DATA_BASE}/portfolios/${portfolioId}/refresh-prices`,
    { method: "POST" },
  );
}

export function refreshAllPortfolioPrices(): Promise<RefreshAllPricesResponse> {
  return apiRequest<RefreshAllPricesResponse>(`${MARKET_DATA_BASE}/refresh-all-prices`, {
    method: "POST",
  });
}

export function searchSymbols(query: string): Promise<SymbolSearchResult[]> {
  const params = new URLSearchParams({ query });
  return apiRequest<SymbolSearchResult[]>(
    `${MARKET_DATA_BASE}/search?${params.toString()}`,
    { method: "GET" },
  );
}
